Source/DafnyCore/AST/IndentationFormatter.cs

```diff
@@ -0,0 +1,1875 @@
+  ///
+  ///       // Comment here
+  ///  const b := 1
+  ///
+  /// // One last word
+  ///   }
+  /// ```
+  ///
+  /// 1. The trailing trivia of the token `)` is
+  /// "\n  // Comment\n        // continued\n\n"
+  ///
+  /// That trivia will be processed with the options
+  /// * trailingTrivia == true
+  /// * precededByNewline == false (irrelevant, a trailing trivia is preceded by a token)
+  /// * indentationBefore == "    " (4 spaces)
+  /// * lastIndentation == "    " (irrelevant)
+  /// The method will return:
+  /// "\n    // Comment\n    // continued\n\n"
+  ///
+  /// 2. The leading trivia of the token `const` is
+  /// "      // Comment here\n "
+  /// That trivia will be processed with the options
+  /// * trailingTrivia == false
+  /// * precededByNewline == true
+  /// * indentationBefore == "  " (2 spaces, it's indentation for any comment before the constant)
+  /// * lastIndentation == "  " (2 spaces, it's the indentation in front of the "const" token if it's on its own line)
+  /// The method will return:
+  /// "  // Comment here\n  "
+  ///
+  /// 3. The leading trivia of the token `b``is
+  /// " "
+  /// That trivia will be processed with the options
+  /// * trailingTrivia == false
+  /// * precededByNewline == false
+  /// * indentationBefore == "    "
+  /// * lastIndentation == "    "
+  /// However, because there are no newlines and it's not preceded
+  /// by a newline, the method will return the same:
+  /// " "
+  /// 
+  /// 4. The leading trivia of the token `}` is
+  /// "// One last word\n"
+  /// That trivia will be processed with the options
+  /// * trailingTrivia == false
+  /// * precededByNewline == true
+  /// * indentationBefore == "  "
+  /// * lastIndentation == ""
+  /// This is an example where a comment before this token should not
+  /// be aligned with that token, the method will return:
+  /// "  // One last word\n"
+  ///
+  /// After all these formattings, the final result will be the expected one
+  /// ```
+  /// module X {
+  ///   datatype A =
+  ///     C()
+  ///     // Comment
+  ///     // continued
+  ///
+  ///   // Comment here
+  ///   const b := 1
+  ///
+  ///   // One last word
+  /// }
+  /// ```
+  public string Reindent(IToken token, bool trailingTrivia, bool precededByNewline, string indentationBefore, string lastIndentation) {
+    var input = trailingTrivia ? token.TrailingTrivia : token.LeadingTrivia;
+    var commentExtra = "";
+    // Invariant: Relative indentation inside a multi-line comment should be unchanged
+    var originalCommentIndent = 0;
+    var newCommentIndent = 0;
+    var previousMatchWasSingleLineCommentToAlign = false;
+
+    // Apply the given rules on a match of a (newline|beginning) + space + comment?
+    string ApplyIndentRules(System.Text.RegularExpressions.Match match) {
+      if (match.Groups["trailingWhitespace"].Success) {
+        return HelperString.RemoveTrailingWhitespace ? "" : match.Groups["trailingWhitespace"].Value;
```

Comment:
> Can we avoid using mutable static state?

I understand that it's better for thread safety. For now, I'll consider cases one by one because I use static fields as a way to interface simply with Dafny, without having to provide explicit options.

For this case, I made `RemoveTrailingWhitespace` readonly.